### PR TITLE
Make KeyBytes allocates less memory

### PR DESCRIPTION
**Pull request by cuonglm**
Closes: #417

If the segments is less than the layouts, the we don't need to allocate
full KeyFormat.length, just allocate enough memory to store the
segments.

```
name                                                old time/op    new time/op    delta
KeyFormat_KeyBytesOneSegment-8                        19.2ns ± 0%    16.4ns ± 1%  -14.62%  (p=0.000 n=9+10)
KeyFormat_KeyBytesThreeSegment-8                      25.6ns ± 0%    25.9ns ± 1%   +1.37%  (p=0.000 n=10+10)
KeyFormat_KeyBytesOneSegmentWithVariousLayouts-8      22.2ns ± 1%    16.4ns ± 0%  -26.14%  (p=0.000 n=10+8)
KeyFormat_KeyBytesThreeSegmentWithVariousLayouts-8    28.8ns ± 3%    28.7ns ± 0%     ~     (p=0.588 n=10+9)

name                                                old alloc/op   new alloc/op   delta
KeyFormat_KeyBytesOneSegment-8                         32.0B ± 0%     16.0B ± 0%  -50.00%  (p=0.000 n=10+10)
KeyFormat_KeyBytesThreeSegment-8                       32.0B ± 0%     32.0B ± 0%     ~     (all equal)
KeyFormat_KeyBytesOneSegmentWithVariousLayouts-8       64.0B ± 0%     16.0B ± 0%  -75.00%  (p=0.000 n=10+10)
KeyFormat_KeyBytesThreeSegmentWithVariousLayouts-8     64.0B ± 0%     64.0B ± 0%     ~     (all equal)

name                                                old allocs/op  new allocs/op  delta
KeyFormat_KeyBytesOneSegment-8                          1.00 ± 0%      1.00 ± 0%     ~     (all equal)
KeyFormat_KeyBytesThreeSegment-8                        1.00 ± 0%      1.00 ± 0%     ~     (all equal)
KeyFormat_KeyBytesOneSegmentWithVariousLayouts-8        1.00 ± 0%      1.00 ± 0%     ~     (all equal)
KeyFormat_KeyBytesThreeSegmentWithVariousLayouts-8      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
```

### CODE REVIEW

**Comment by cuonglm** [comment]
cc @odeke-em

**Review by odeke-em — APPROVED** [review]
LGTM, thank you @cuonglm! Please just add a comment right above to provide reasoning why.

File: key_format.go
Comment (cuonglm) [reply key_format.go:45]: Done.

**Review by robert-zaremba — APPROVED** [review]

**Comment by cuonglm** [comment]
@odeke-em @robert-zaremba somehow the test haven't run, cause automerge failed

**Comment by robert-zaremba** [comment]
I've just re run it.

**Comment by cuonglm** [comment]
cc @odeke-em @robert-zaremba for re-running the workflow.

**Comment by odeke-em** [comment]
@robert-zaremba thank you for the merge, please later on help cut a new release so that the cosmos-sdk and gaiad can benefit from this change as it is heavily used

**Comment by cuonglm** [comment]
@odeke-em maybe wait until all the PRs merged. There is #420 and another upcoming PR for encodeVarUint after #420 merged

**Pull request merged**
